File: atlan-performance-mobile/iosApp/AtlanPerformance/Design/AtlanComponents.swift
import SwiftUI
import Shared

/// Atlan custom SwiftUI components. No component introduces streak / rank / flame / trophy / badge
/// pressure. Coral is reserved for productive action; Tide marks Why affordances.
///
/// Pattern coverage (atlan-mobile-design-patterns.md): every interactive control exposes a pressed
/// state (§9.2), meets the 44pt minimum touch target (§11.2), and never signals state through colour
/// alone (§11.5). Async actions get explicit loading + disabled states (§8.1, §10.5–§10.6).

/// Minimum touch target for any tappable control (§11.2 — iOS 44pt).
private let atlanMinTouchTarget: CGFloat = 44

/// Shared pressed-state treatment: a calm opacity + scale shift, no layout jitter (§9.2).
struct AtlanPressStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .opacity(configuration.isPressed ? 0.82 : 1.0)
            .scaleEffect(configuration.isPressed ? 0.98 : 1.0)
            .animation(.easeOut(duration: 0.12), value: configuration.isPressed)
    }
}

/// Primary CTA. Abyss pill by default; Coral only for productive action (e.g. start / complete).
/// Supports disabled (§10.5) and loading (§10.6 — spinner, preserved width, no double-submit) states.
struct AtlanButton: View {
    let title: String
    var coral: Bool = false
    var isEnabled: Bool = true
    var isLoading: Bool = false
    let action: () -> Void

    private var fill: Color {
        isEnabled ? (coral ? AtlanColors.coralBright : AtlanColors.abyss) : AtlanColors.tideSoft
    }

    var body: some View {
        Button(action: { if isEnabled && !isLoading { action() } }) {
            ZStack {
                // Keep the label in the layout (hidden) while loading so width never shifts (§9.4).
                Text(title)
                    .font(.headline)
                    .opacity(isLoading ? 0 : 1)
                if isLoading {
                    ProgressView().tint(AtlanColors.foam)
                }
            }
            .foregroundColor(AtlanColors.foam)
            .frame(maxWidth: .infinity, minHeight: atlanMinTouchTarget)
            .padding(.vertical, AtlanSpacing.md)
            .background(fill)
            .clipShape(Capsule())
        }
        .buttonStyle(AtlanPressStyle())
        .disabled(!isEnabled || isLoading)
        .accessibilityLabel(title)
        .accessibilityValue(isLoading ? "Loading" : "")
        .accessibilityHint(isEnabled ? "" : "Unavailable")
    }
}

/// Small status/label pill.
struct AtlanPill: View {
    let text: String
    var body: some View {
        Text(text)
            .font(.caption).bold()
            .foregroundColor(AtlanColors.tideDeep)
            .padding(.horizontal, AtlanSpacing.md)
            .padding(.vertical, AtlanSpacing.xs)
            .background(AtlanColors.tidePale)
            .clipShape(Capsule())
    }
}

/// Tide-outlined circular Why affordance with an italic i. Tide — never Coral. The visible ring stays
/// 24pt for calm density, but the tap target expands to the 44pt minimum (§11.2).
struct AtlanInfoButton: View {
    var accessibilityLabel: String = "Why — open explanation"
    let action: () -> Void
    var body: some View {
        Button(action: action) {
            Text("i")
                .italic().bold()
                .foregroundColor(AtlanColors.tide)
                .frame(width: 24, height: 24)
                .overlay(Circle().stroke(AtlanColors.tide, lineWidth: 1.5))
                .frame(width: atlanMinTouchTarget, height: atlanMinTouchTarget)
                .contentShape(Rectangle())
        }
        .buttonStyle(AtlanPressStyle())
        .accessibilityLabel(accessibilityLabel)
        .accessibilityAddTraits(.isButton)
    }
}

/// Single-select option row (Calibration, language-style choices). Selection is signalled by a
/// checkmark + border + fill, never colour alone (§11.5), and exposes the `.isSelected` trait so
/// VoiceOver announces state (§11.4). Whole row is tappable and meets the 44pt minimum (§8.2, §11.2).
struct AtlanSelectableRow: View {
    let title: String
    let isSelected: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            HStack(spacing: AtlanSpacing.md) {
                Text(title)
                    .font(AtlanTypography.body)
                    .foregroundColor(isSelected ? AtlanColors.tideDeep : AtlanColors.abyss)
                    .fixedSize(horizontal: false, vertical: true)
                    .frame(maxWidth: .infinity, alignment: .leading)
                Image(systemName: isSelected ? "checkmark.circle.fill" : "circle")
                    .foregroundColor(isSelected ? AtlanColors.tide : AtlanColors.tideSoft)
            }
            .padding(AtlanSpacing.lg)
            .frame(minHeight: atlanMinTouchTarget)
            .background(isSelected ? AtlanColors.tidePale : AtlanColors.paper)
            .overlay(
                RoundedRectangle(cornerRadius: AtlanRadii.lg)
                    .stroke(isSelected ? AtlanColors.tide : AtlanColors.tideSoft,
                            lineWidth: isSelected ? 1.5 : 1)
            )
            .clipShape(RoundedRectangle(cornerRadius: AtlanRadii.lg))
        }
        .buttonStyle(AtlanPressStyle())
        .accessibilityAddTraits(isSelected ? [.isButton, .isSelected] : .isButton)
    }
}

/// Dashboard metric chip with inline Why affordance.
struct AtlanMetricChip: View {
    let chip: MetricChip
    let onWhy: (String) -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: AtlanSpacing.xs) {
            HStack {
                Text(chip.title.uppercased())
                    .font(AtlanTypography.label).foregroundColor(AtlanColors.tideDeep)
                if let key = chip.whyConceptKey {
                    AtlanInfoButton(accessibilityLabel: "Why \(chip.title) — open explanation") { onWhy(key) }
                }
            }
            Text(chip.value).font(AtlanTypography.numeric).foregroundColor(AtlanColors.abyss)
            Text(chip.detail).font(AtlanTypography.body).foregroundColor(AtlanColors.abyss)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(AtlanSpacing.lg)
        .background(AtlanColors.paper)
        .clipShape(RoundedRectangle(cornerRadius: AtlanRadii.lg))
    }
}

/// Reusable calm error surface (§10.8). No blame, no red — Tide retry icon, explains the issue and
/// reassures that data is safe, offers Retry (and an optional safe exit). Used as a load-failure
/// fallback so unrecoverable states never strand the user.
struct AtlanErrorView: View {
    let title: String
    let message: String
    var retryTitle: String = "Retry"
    let onRetry: () -> Void
    var exitTitle: String? = nil
    var onExit: (() -> Void)? = nil

    var body: some View {
        VStack(spacing: AtlanSpacing.md) {
            Image(systemName: "arrow.clockwise.circle")
                .font(.system(size: 44)).foregroundColor(AtlanColors.tideDeep)
            Text(title).font(AtlanTypography.title).foregroundColor(AtlanColors.abyss)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)
            Text(message).foregroundColor(AtlanColors.tideDeep)
                .multilineTextAlignment(.center)
                .fixedSize(horizontal: false, vertical: true)
            AtlanButton(title: retryTitle) { onRetry() }
                .padding(.top, AtlanSpacing.sm)
            if let onExit = onExit, let exitTitle = exitTitle {
                Button { onExit() } label: {
                    Text(exitTitle).foregroundColor(AtlanColors.tideDeep)
                        .frame(maxWidth: .infinity, minHeight: 44).contentShape(Rectangle())
                }
                .buttonStyle(AtlanPressStyle())
            }
        }
        .frame(maxWidth: .infinity)
        .padding(.vertical, AtlanSpacing.xxxl)
    }
}

/// One half of the Wet Mode bottom action grid. Large touch zone, no precision tap. Coral only for
/// the productive Complete zone. Exposes an accessibility label so swipe is never the only path.
struct WetModeActionZone: View {
    let title: String
    let hint: String
    let accessibleLabel: String
    let productive: Bool
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            VStack(spacing: AtlanSpacing.xs) {
                Text(title).font(.system(size: 28, weight: .bold)).foregroundColor(AtlanColors.foam)
                Text(hint).foregroundColor(AtlanColors.foam)
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity)
            .background(productive ? AtlanColors.coralBright : AtlanColors.abyss)
        }
        .buttonStyle(AtlanPressStyle())
        .accessibilityLabel(accessibleLabel)
        .accessibilityAddTraits(.isButton)
    }
}

File: atlan-performance-mobile/androidApp/src/main/java/com/atlan/performance/android/design/AtlanComponents.kt
package com.atlan.performance.android.design

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.scale
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.atlan.performance.shared.design.AtlanRadii
import com.atlan.performance.shared.design.AtlanSpacing
import com.atlan.performance.shared.presentation.MetricChip

/**
 * Atlan custom component library. Built on Compose primitives, not Material's default visual
 * language. No component introduces streak / rank / flame / trophy / badge pressure.
 *
 * Pattern coverage (atlan-mobile-design-patterns.md): every interactive control exposes a pressed
 * state (§9.2), meets the 48dp minimum touch target (§11.2), and never signals state through colour
 * alone (§11.5). Async actions get explicit loading + disabled states (§8.1, §10.5–§10.6).
 */

/** Minimum touch target for any tappable control (§11.2 — Android 48dp). */
private val AtlanMinTouchTarget = 48.dp

/** Primary CTA. Dark Abyss pill by default; Coral only where productive action warrants it. */
@Composable
fun AtlanButton(
    text: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    coral: Boolean = false,
    enabled: Boolean = true,
    loading: Boolean = false
) {
    val active = enabled && !loading
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val scale by animateFloatAsState(if (pressed && active) 0.98f else 1f, label = "atlanButtonPress")
    val bg = when {
        !enabled -> AtlanPalette.TideSoft
        coral -> AtlanPalette.CoralBright
        else -> AtlanPalette.Abyss
    }
    Row(
        modifier = modifier
            .fillMaxWidth()
            .scale(scale)
            .clip(RoundedCornerShape(AtlanRadii.pill.dp))
            .background(bg)
            .heightIn(min = AtlanMinTouchTarget)
            .alpha(if (pressed && active) 0.82f else 1f)
            .clickable(
                interactionSource = interaction,
                indication = null,
                enabled = active,
                onClick = onClick
            )
            .semantics { contentDescription = text; role = Role.Button }
            .padding(vertical = AtlanSpacing.md.dp, horizontal = AtlanSpacing.xl.dp),
        horizontalArrangement = Arrangement.Center,
        verticalAlignment = Alignment.CenterVertically
    ) {
        if (loading) {
            // Preserve height; never let layout shift while async work runs (§9.4).
            CircularProgressIndicator(
                color = AtlanPalette.Foam,
                strokeWidth = 2.dp,
                modifier = Modifier.size(20.dp)
            )
        } else {
            Text(text, color = AtlanPalette.Foam, fontWeight = FontWeight.SemiBold, fontSize = 16.sp)
        }
    }
}

/**
 * Back affordance for pushed screens. Replaces a bare "←" glyph so the tap target meets the 48dp
 * minimum (§11.2) and carries a localized content description + button role for TalkBack (§11.4).
 */
@Composable
fun AtlanBackButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    contentDescription: String = "Back"
) {
    Box(
        modifier = modifier
            .size(AtlanMinTouchTarget)
            .clip(RoundedCornerShape(AtlanRadii.pill.dp))
            .clickable(onClick = onClick)
            .semantics { this.contentDescription = contentDescription; role = Role.Button },
        contentAlignment = Alignment.CenterStart
    ) {
        Text("‹", color = AtlanPalette.Abyss, fontSize = 30.sp, fontWeight = FontWeight.Bold)
    }
}

/** Small status/label pill (e.g. progress "2 of 4", "Offline · Cached"). */
@Composable
fun AtlanPill(text: String, modifier: Modifier = Modifier) {
    Row(
        modifier = modifier
            .clip(RoundedCornerShape(AtlanRadii.pill.dp))
            .background(AtlanPalette.TidePale)
            .padding(horizontal = AtlanSpacing.md.dp, vertical = AtlanSpacing.xs.dp)
    ) {
        Text(text, color = AtlanPalette.TideDeep, fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
}

/**
 * Tide-outlined circular "Why" affordance with an italic i. Tide — never Coral. The visible ring
 * stays 24dp for calm density, but the tap target expands to the 48dp minimum (§11.2).
 */
@Composable
fun AtlanInfoButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    accessibleLabel: String = "Why — open explanation"
) {
    Box(
        modifier = modifier
            .size(AtlanMinTouchTarget)
            .clip(RoundedCornerShape(AtlanRadii.pill.dp))
            .clickable(onClick = onClick)
            .semantics { contentDescription = accessibleLabel; role = Role.Button },
        contentAlignment = Alignment.Center
    ) {
        Box(
            modifier = Modifier
                .size(24.dp)
                .clip(RoundedCornerShape(AtlanRadii.pill.dp))
                .border(1.5.dp, AtlanPalette.Tide, RoundedCornerShape(AtlanRadii.pill.dp)),
            contentAlignment = Alignment.Center
        ) {
            Text("i", color = AtlanPalette.Tide, fontStyle = FontStyle.Italic, fontWeight = FontWeight.Bold)
        }
    }
}

/**
 * Single-select option row (Calibration, language-style choices). Selection is signalled by a
 * check glyph + border + fill, never colour alone (§11.5), and exposes the selected semantics state
 * so TalkBack announces it (§11.4). Whole row is tappable and meets the 48dp minimum (§8.2, §11.2).
 */
@Composable
fun AtlanSelectableRow(
    title: String,
    isSelected: Boolean,
    onClick: () -> Unit,
    modifier: Modifier = Modifier
) {
    val borderColor = if (isSelected) AtlanPalette.Tide else AtlanPalette.TideSoft
    val bg = if (isSelected) AtlanPalette.TidePale else AtlanPalette.Paper
    Row(
        modifier = modifier
            .fillMaxWidth()
            .heightIn(min = AtlanMinTouchTarget)
            .clip(RoundedCornerShape(AtlanRadii.lg.dp))
            .background(bg)
            .border(if (isSelected) 1.5.dp else 1.dp, borderColor, RoundedCornerShape(AtlanRadii.lg.dp))
            .clickable(onClick = onClick)
            .semantics { selected = isSelected; role = Role.RadioButton }
            .padding(AtlanSpacing.lg.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Text(
            title,
            color = if (isSelected) AtlanPalette.TideDeep else AtlanPalette.Abyss,
            style = AtlanType.Body,
            modifier = Modifier.weight(1f)
        )
        Text(
            if (isSelected) "✓" else "○",
            color = if (isSelected) AtlanPalette.Tide else AtlanPalette.TideSoft,
            fontWeight = FontWeight.Bold,
            fontSize = 18.sp
        )
    }
}

/** Dashboard metric chip with inline Why affordance. */
@Composable
fun AtlanMetricChip(chip: MetricChip, onWhy: (String) -> Unit, modifier: Modifier = Modifier) {
    Column(
        modifier = modifier
            .clip(RoundedCornerShape(AtlanRadii.lg.dp))
            .background(AtlanPalette.Paper)
            .padding(AtlanSpacing.lg.dp)
    ) {
        Row(verticalAlignment = Alignment.CenterVertically) {
            Text(chip.title.uppercase(), style = AtlanType.Label, color = AtlanPalette.TideDeep)
            chip.whyConceptKey?.let {
                AtlanInfoButton(
                    onClick = { onWhy(it) },
                    accessibleLabel = "Why ${chip.title} — open explanation"
                )
            }
        }
        Text(chip.value, style = AtlanType.Numeric, color = AtlanPalette.Abyss)
        Text(chip.detail, style = AtlanType.Body, color = AtlanPalette.Abyss)
    }
}

/**
 * Reusable calm error surface (§10.8). No blame, no red — Tide retry glyph, explains the issue and
 * reassures that data is safe, offers Retry (and an optional safe exit). Used as a load-failure
 * fallback so unrecoverable states never strand the user.
 */
@Composable
fun AtlanErrorScreen(
    title: String,
    message: String,
    onRetry: () -> Unit,
    modifier: Modifier = Modifier,
    retryText: String = "Retry",
    exitText: String? = null,
    onExit: (() -> Unit)? = null
) {
    Column(
        modifier = modifier.fillMaxWidth().padding(AtlanSpacing.xl.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(AtlanSpacing.md.dp)
    ) {
        Text("↻", color = AtlanPalette.TideDeep, fontSize = 44.sp)
        Text(title, style = AtlanType.Display, color = AtlanPalette.Abyss, textAlign = TextAlign.Center)
        Text(message, color = AtlanPalette.TideDeep, textAlign = TextAlign.Center)
        AtlanButton(text = retryText, onClick = onRetry, modifier = Modifier.padding(top = AtlanSpacing.sm.dp))
        if (onExit != null && exitText != null) {
            Text(
                exitText,
                modifier = Modifier
                    .fillMaxWidth()
                    .heightIn(min = 48.dp)
                    .clickable(onClick = onExit)
                    .padding(AtlanSpacing.sm.dp),
                color = AtlanPalette.TideDeep,
                textAlign = TextAlign.Center
            )
        }
    }
}

/** Card wrapper for session content. */
@Composable
fun AtlanSessionCard(modifier: Modifier = Modifier, content: @Composable androidx.compose.foundation.layout.ColumnScope.() -> Unit) {
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(AtlanRadii.phone.dp))
            .background(AtlanPalette.Abyss)
            .padding(AtlanSpacing.xl.dp),
        verticalArrangement = Arrangement.spacedBy(AtlanSpacing.sm.dp),
        content = content
    )
}

/** Native bottom sheet wrapper for Why Modal and Session Swapper. */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun AtlanBottomSheet(onDismiss: () -> Unit, content: @Composable () -> Unit) {
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    ModalBottomSheet(
        onDismissRequest = onDismiss,
        sheetState = sheetState,
        containerColor = AtlanPalette.Paper,
        shape = RoundedCornerShape(topStart = AtlanRadii.sheet.dp, topEnd = AtlanRadii.sheet.dp)
    ) {
        Column(Modifier.padding(AtlanSpacing.xl.dp)) { content() }
    }
}

/**
 * One half of the Wet Mode bottom action grid. Large touch zone, no precision tap. Exposes an
 * accessible action so swipe is never the only path. Coral only for the Complete (productive) zone.
 */
@Composable
fun WetModeActionZone(
    title: String,
    hint: String,
    accessibleLabel: String,
    productive: Boolean,
    onActivate: () -> Unit,
    modifier: Modifier = Modifier
) {
    val bg = if (productive) AtlanPalette.CoralBright else AtlanPalette.Abyss
    Column(
        modifier = modifier
            .background(bg)
            .clickable(onClick = onActivate)
            .semantics { contentDescription = accessibleLabel; role = Role.Button }
            .padding(AtlanSpacing.xl.dp),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.Center
    ) {
        Text(title, color = AtlanPalette.Foam, fontSize = 28.sp, fontWeight = FontWeight.Bold)
        Text(hint, color = AtlanPalette.Foam, fontSize = 14.sp, textAlign = TextAlign.Center)
    }
}
